### ISSUE TITLE
Stack overflow when validating AST with great height

### ISSUE THREAD

**Issue by mor-n4** [eclipse/xtext-core#449]
When having an AST with a great height (e.g. deeply nested expressions, long chains of method calls when using fluent APIs) the Xtext validator can run into a stack overflow, see stack trace below.

To reproduce:

1. create new empty Xtext project (with wizard)
2. use the following grammar:
    ```
    Script:
        (exprs+=Expression ';')*;
    
    Expression:
        Literal (
            {CallExpression.target=current} '(' args+=Literal (',' args+=Literal)* ')'
        |    {AccessExpression.target=current} '.' memberName=ID
        )*;
    
    Literal:
        {StringLiteral} value=STRING;
    ```
3. open the following sample file in an editor:
    ```
    'hello'.append('foo')
        .append('foo')
        .append('foo')
        .append('foo');
    ```
4. now duplicate line 2 or 3 until you have around 600 lines (on my system; number of lines required to provoke stack overflow may depend on system configuration)

Note that it is not required to add any custom validations.

Here's the stack trace that I got:
```
java.lang.StackOverflowError
    at org.apache.log4j.Category.forcedLog(Category.java:391)
    at org.apache.log4j.Category.error(Category.java:322)
    at org.eclipse.xtext.validation.CompositeEValidator.validate(CompositeEValidator.java:155)
    at org.eclipse.emf.ecore.util.Diagnostician.doValidate(Diagnostician.java:171)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:158)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:137)
    at org.eclipse.xtext.validation.CancelableDiagnostician.validate(CancelableDiagnostician.java:39)
    at org.eclipse.emf.ecore.util.Diagnostician.doValidateContents(Diagnostician.java:181)
    at org.eclipse.xtext.validation.CancelableDiagnostician.doValidateContents(CancelableDiagnostician.java:70)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:161)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:137)
    at org.eclipse.xtext.validation.CancelableDiagnostician.validate(CancelableDiagnostician.java:39)
    at org.eclipse.emf.ecore.util.Diagnostician.doValidateContents(Diagnostician.java:181)
    at org.eclipse.xtext.validation.CancelableDiagnostician.doValidateContents(CancelableDiagnostician.java:70)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:161)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:137)
    at org.eclipse.xtext.validation.CancelableDiagnostician.validate(CancelableDiagnostician.java:39)
    at org.eclipse.emf.ecore.util.Diagnostician.doValidateContents(Diagnostician.java:181)
    at org.eclipse.xtext.validation.CancelableDiagnostician.doValidateContents(CancelableDiagnostician.java:70)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:161)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:137)
    at org.eclipse.xtext.validation.CancelableDiagnostician.validate(CancelableDiagnostician.java:39)
    at org.eclipse.emf.ecore.util.Diagnostician.doValidateContents(Diagnostician.java:181)
    at org.eclipse.xtext.validation.CancelableDiagnostician.doValidateContents(CancelableDiagnostician.java:70)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:161)
    at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:137)
    at org.eclipse.xtext.validation.CancelableDiagnostician.validate(CancelableDiagnostician.java:39)
    at org.eclipse.emf.ecore.util.Diagnostician.doValidateContents(Diagnostician.java:181)
    at org.eclipse.xtext.validation.CancelableDiagnostician.doValidateContents(CancelableDiagnostician.java:70)
(goes on ...)
```

**Comment by cdietrich** [comment]
Workaround: increase stack size

**Comment by cdietrich** [comment]
i wonder if this is not a bug in emf. CancelableDiagnostician.doValidateContents calls super. the other classes are in emf. @merks what do you think?

**Comment by cdietrich** [comment]
sample ecore

```
<?xml version="1.0" encoding="UTF-8"?>
<ecore:EPackage xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:ecore="http://www.eclipse.org/emf/2002/Ecore" name="sample" nsURI="http://www.itemis.de/sample" nsPrefix="sample">
  <eClassifiers xsi:type="ecore:EClass" name="Node">
    <eStructuralFeatures xsi:type="ecore:EReference" name="children" upperBound="-1"
        eType="#//Node" containment="true"/>
  </eClassifiers>
</ecore:EPackage>
```

and main to reproduce in pure emf

```
import org.eclipse.emf.ecore.util.Diagnostician;

import sample.Node;
import sample.SampleFactory;

public class Main {
	public static void main(String[] args) {
		Node root = SampleFactory.eINSTANCE.createNode();
		Node current = root;
		for (int i = 0; i< 10000; i++) {
			Node node = SampleFactory.eINSTANCE.createNode();
			current.getChildren().add(node);
			current = node;
		}
		Diagnostician.INSTANCE.validate(root);
	}

}

```

**Comment by merks** [comment]
Of course many algorithms are recursive and if you make the recursion deep enough, the stack will overflow.  When validating the contents of some object, those children (contained objects) might well be from a different Ecore model with a different validator so we must delegate back to the root validator which includes the logic for finding the right per-package validator.

**Comment by cdietrich** [comment]
so you would prefer a clone wontfix?

**Comment by merks** [comment]
Well, in general any recursion can suffer such a problem given an input that induces deep enough recursion.  One would need to redesgin the algorithm to avoid recursion.  That's not possible in this case without breaking derived classes that override doValidateContents, i.e., that are specialized in a way that relies on the recursion.  But perhaps a creative solution is possible. I could imagine maintaining a list of contents that need validating, and perhaps using reflection to test if the current class is a derived class with an override of doCreateContents.  Note that org.eclipse.xtext.xbase.validation.XbaseDiagnostician.doValidateContents(EObject, DiagnosticChain, Map<Object, Object>) is such a case, though the class Javadoc says "This is for EMF 2.5 compatibility"  but that's such an incredibly old version it's hard to imagine that being important. Please open an EMF bugzilla and I'll investigate further, though in the end I might close that as wontfix.

**Comment by cdietrich** [comment]
created https://bugs.eclipse.org/bugs/show_bug.cgi?id=521642 and thus will close this one

**Comment by cdietrich** [comment]
needs testing with the changed emf version https://bugs.eclipse.org/bugs/show_bug.cgi?id=521642

**Comment by cdietrich** [comment]
@kthoms @szarnekow who of you could evaluate this. am still quite busy

**Comment by szarnekow** [comment]
@cdietrich  I did not find the time to look into this yet.

**Comment by knutwannheden** [comment]
Reminds me of a similar issue I had with EContentAdapters: https://www.eclipse.org/forums/index.php/t/781423/. The JIT also contributed its share to the stack size.

**Comment by cdietrich** [comment]
it would still be nice if someone could have a look

**Comment by mor-n4** [comment]
I did some tests with the new EMF version provided by Ed.

1) With Xtext Complete SDK 2.12.0.v20170519-1412 and EMF 2.13.0.v20170609-0707 (official release) I was able to reproduce the stack trace from the issue description above.

2) Then, I upgraded "EMF - Eclipse Modeling Framework Core Runtime" and "EMF Common" to version 2.14.0.v20180128-0931 (located at https://ci.eclipse.org/emf/job/gerrit/4/artifact/updates) but did not change anything else, in particular Xtext version is still 2.12.0. The original stack overflow did not appear again. However a new stack overflow with a different stack trace showed up [1].

3) Also tested with Xtext 2.13.0, but wasn't able to reproduce the original stack overflow, even when using an official release of EMF. I always got the overflow with the "new" stack trace [1].

Overall, it seems the EMF version at https://ci.eclipse.org/emf/job/gerrit/4/artifact/updates fixes the original problem.

Hope this helps, though I can't comment on the Xtext internals mentioned by Ed.

[1] "new" stack trace:
```
java.lang.StackOverflowError
	at org.eclipse.emf.ecore.util.EContentAdapter.selfAdapt(EContentAdapter.java:54)
	at org.eclipse.emf.ecore.util.EContentAdapter.notifyChanged(EContentAdapter.java:43)
	at org.eclipse.xtext.util.OnChangeEvictingCache$CacheAdapter.notifyChanged(OnChangeEvictingCache.java:269)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:523)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:1)
	at org.eclipse.emf.common.util.ArrayDelegatingEList.remove(ArrayDelegatingEList.java:656)
	at org.eclipse.emf.common.util.AbstractEList.remove(AbstractEList.java:456)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:388)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:383)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:317)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:288)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:267)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:527)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:1)
	at org.eclipse.emf.common.util.ArrayDelegatingEList.remove(ArrayDelegatingEList.java:656)
	at org.eclipse.emf.common.util.AbstractEList.remove(AbstractEList.java:456)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:388)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:383)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:317)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:288)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:267)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:527)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:1)
	at org.eclipse.emf.common.util.ArrayDelegatingEList.remove(ArrayDelegatingEList.java:656)
	at org.eclipse.emf.common.util.AbstractEList.remove(AbstractEList.java:456)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:388)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:383)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:317)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:288)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:267)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:527)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:1)
	at org.eclipse.emf.common.util.ArrayDelegatingEList.remove(ArrayDelegatingEList.java:656)
	at org.eclipse.emf.common.util.AbstractEList.remove(AbstractEList.java:456)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:388)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:383)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:317)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:288)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:267)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:527)
	at org.eclipse.emf.ecore.impl.MinimalEObjectImpl$1ArrayDelegatingAdapterList.didRemove(MinimalEObjectImpl.java:1)
	at org.eclipse.emf.common.util.ArrayDelegatingEList.remove(ArrayDelegatingEList.java:656)
	at org.eclipse.emf.common.util.AbstractEList.remove(AbstractEList.java:456)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:388)
	at org.eclipse.emf.ecore.util.EContentAdapter.removeAdapter(EContentAdapter.java:383)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:317)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:288)
	at org.eclipse.emf.ecore.util.EContentAdapter.unsetTarget(EContentAdapter.java:267)
...
```

**Comment by cdietrich** [comment]
cc @merks do you already know the new Stacktrace?

**Comment by cdietrich** [comment]
@mor-n4 do you have the begin of the stackoverflow too?

**Comment by ewillink** [comment]
Ed M has just gone on holiday.

The EContentAdapter stack trace does not appear to involve any recently changed code, so I suspect that it is either a new phenomenon or bad user code.

Looks like the kind of problem that is easy to fix with a repro, impossible to guess at without.

I suggest raising a Bugzilla with a small repro.

**Comment by knutwannheden** [comment]
Now this does sound very similar to the problem I had some years ago, although I ran into this when adding an EContentAdapter and here it seems to happen when it is removed.

IMHO increasing the stack size should not only be considered a workaround, as I think there may be many other pieces of code which may end up failing for the same reason. Yet I think it would still make sense to analyze the individual cases to see if the code can be changed to use a non-recursive algorithm or otherwise refactored to decrease the stack size.

**Comment by cdietrich** [comment]
sounds like a different issue
can you create a new xtext bug for that?

i wonder if there is something special about your metamodel

**Comment by cdietrich** [comment]
maybe this is a similar effect as described in https://github.com/eclipse/xtext-core/issues/122

**Comment by cdietrich** [comment]
[UpdateSite.zip](https://github.com/eclipse/xtext-core/files/1758616/UpdateSite.zip)

**Comment by merks** [comment]
Note that there is a fix for the EContentAdapter stack overflow as well:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=532482

Xtext will need to make changes to its derived EContentAdapters to take advantage of this.

Fixes for both the Diagnostician and the EContentAdapter stack overflows are available in nightly builds:

http://download.eclipse.org/modeling/emf/emf/builds/nightly/latest

**Comment by cdietrich** [comment]
fixed in 2.14
thx @ArneDeutsch  and @szarnekow
